### PR TITLE
Bug #1128005 - Emphasize the downbeats of a track by a separate marker on the waveform

### PR DESCRIPTION
**Pull request by ghost**
This PR attempts a simple solution to bug #1128005. It introduces measure (downbeat, every 4 beats) and phrase (16 beats) markers as part of the Waveform view. The location of these markers will eventually be determined by the Cue or a new Cue type, especially designated for beat grid anchoring. The main purpose of this feature is to facilitate better beat-matching. 

The implementation plan for this feature is:
1. Add measure and phrase markers (every 4 and 16 beats, number of beats and color hard coded), the first marker is the same as the first beat of the track
2. Make the number of beats per measure/phrase and colors part of Preferences/mixxx.cfg 
3. Change the logic so CUE (Cue::LOAD) ) determines the first beat
4. Add a new cue type (beat grid anchor or first beat) to serve the same role, restore CUE functionality

mixxx_beatgrid_downbeat2
mixxx_beatgrid_downbeat3

### CODE REVIEW

**Comment by ritschwumm** [comment]
just a few works about my experiences with using something similar:

4 measures per phrase works quite well for many tracks, but there are a surprising number of them where the scheme _shifts_ by 2 measures somewhere in the middle. quite irritating, when you're mixing.
whenever this happens, i wish i could either tell my app such a shift happenend, or at least use 2 measures per phrase for this specific track.

**Comment by nttmusic** [comment]
You are 100% right, this mirrors my experience as well. However, with this one, I'd like to take more gradual approach first (with the simples steps outlined in the overview). Hopefully at some point we can introduce multiple "beat cues" that can also be used for:
1.tempo changes (instant or gradual)
2.meter changes (4/4, 3/4, etc)
3.Phrase length changes

File: src/waveform/renderers/waveformrenderbeat.cpp [hunk 1]
```diff
@@ -83,11 +87,29 @@ void WaveformRenderBeat::draw(QPainter* painter, QPaintEvent* /*event*/) {
...
         xBeatPoint = qRound(xBeatPoint);
 
+        int beatIndex = it->beatIndex();
+
         // If we don't have enough space, double the size.
         if (beatCount >= m_beats.size()) {
             m_beats.resize(m_beats.size() * 2);
         }
 
+        if(beatIndex >= 0) {
```
Comment (daschuer) [inline src/waveform/renderers/waveformrenderbeat.cpp]: This logic needs to be moved at least to the 
Iterator. 
I suggest a isBar() and isPhraseStart()
A phrase does not necessarily strat with a bar.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you for this PR, it is a good start.
But let's hold it up until we can handle other than 4/4 measure.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you for this PR, it is a good start.
But let's hold it up until we can handle other than 4/4 measure.

**Comment by christophski** [comment]
I use Mixxx mainly for funk & disco, you'd be surprised how many weird things go on in these tracks. This is a really nice idea but definitely needs more work. Good job so far. Even if it was something like "section markers", or just various types of markers other than Cues, that might be helpful!

**Comment by nttmusic** [comment]
The requested changes are quite easy to add, however I am starting to think we should have Time Signature (2/4, 3/4, 4/4, etc.) added as part of the Track class first.

**Comment by daschuer** [comment]
Yes, right, this should be available in the library column as well. Be aware that we will have the same issue like BPM and key: We can only show one significant (the first?) value in the library, while the actual value may change during the track.

**Comment by daschuer** [comment]
Matroska has a MEASURE Tag: https://matroska.org/technical/specs/tagging/index.html

**Pull request closed without merging**